src/main/java/org/candlepin/subscriptions/marketplace/MarketplaceSubscriptionIdProvider.java

```diff
@@ -0,0 +1,120 @@
+import org.candlepin.subscriptions.files.ProductProfileRegistry;
+import org.candlepin.subscriptions.subscription.SubscriptionSyncController;
+import org.candlepin.subscriptions.tally.UsageCalculation.Key;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+
+import java.time.OffsetDateTime;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Class responsible for searching Swatch database for subscriptionIds corresponding to usage keys and if
+ * none is found, delegates fetching the subscriptionId to the {@link MarketplaceSubscriptionCollector}.
+ */
+@Component
+public class MarketplaceSubscriptionIdProvider {
+    private static final Logger log = LoggerFactory.getLogger(MarketplaceSubscriptionIdProvider.class);
+
+    private final MarketplaceSubscriptionCollector collector;
+    private final SubscriptionRepository subscriptionRepo;
+    private final SubscriptionSyncController syncController;
+    private final ProductProfileRegistry profileRegistry;
+
+    @Autowired
+    public MarketplaceSubscriptionIdProvider(MarketplaceSubscriptionCollector collector,
+        SubscriptionRepository subscriptionRepo, SubscriptionSyncController syncController,
+        ProductProfileRegistry profileRegistry) {
+        this.collector = collector;
+        this.subscriptionRepo = subscriptionRepo;
+        this.syncController = syncController;
+        this.profileRegistry = profileRegistry;
+    }
+
+    public Optional<Object> findSubscriptionId(String accountNumber, Key usageKey,
+        OffsetDateTime rangeStart, OffsetDateTime rangeEnd) {
+        Assert.isTrue(Usage._ANY != usageKey.getUsage(), "Usage cannot be _ANY");
+        Assert.isTrue(ServiceLevel._ANY != usageKey.getSla(), "Service Level cannot be _ANY");
+
+        String productId = usageKey.getProductId();
+        ProductProfile profile = profileRegistry.findProfileForSwatchProductId(productId);
+        Set<String> roles = profile.getRolesBySwatchProduct().getOrDefault(productId, Collections.emptySet());
+
+        List<Subscription> result = fetchSubscriptions(accountNumber, usageKey, roles, rangeStart, rangeEnd);
+
+        if (result.isEmpty()) {
+            /* If we are missing the subscription, call out to the MarketplaceSubscriptionCollector
+               to fetch from Marketplace.  Sync all those subscriptions. Query again. */
+            var subscriptions = collector.fetchSubscription(accountNumber, usageKey);
+            subscriptions.forEach(syncController::syncSubscription);
+            result = fetchSubscriptions(accountNumber, usageKey, roles, rangeStart, rangeEnd);
+        }
+
+        if (result.isEmpty()) {
+            return Optional.empty();
```

Comment:
```suggestion
            return Optional.of(properties.getDummyId());
```
Here is where I would like to see us return the dummy ID instead of populating a dummy subscription.